arangod/Aql/EnumerateCollectionExecutor.cpp

```diff
@@ -254,13 +265,148 @@ std::tuple<ExecutionState, EnumerateCollectionStats, size_t> EnumerateCollection
   return std::make_tuple(ExecutionState::HASMORE, stats, actuallySkipped);  // tuple, cannot use initializer list due to build failure
 }
 
+uint64_t EnumerateCollectionExecutor::skipEntries(size_t toSkip,
+                                                   EnumerateCollectionStats& stats) {
+  uint64_t actuallySkipped = 0;
+
+  if (_infos.getFilter() == nullptr) {
+    _cursor->skip(toSkip, actuallySkipped);
+    stats.incrScanned(actuallySkipped);
+    _documentProducingFunctionContext.getAndResetNumScanned();
+  } else {
+    _cursor->nextDocument(_documentSkipper, toSkip);
+    size_t filtered = _documentProducingFunctionContext.getAndResetNumFiltered();
+    size_t scanned = _documentProducingFunctionContext.getAndResetNumScanned();
+    TRI_ASSERT(scanned >= filtered);
+    stats.incrFiltered(filtered);
+    stats.incrScanned(scanned);
+    actuallySkipped = scanned - filtered;
+  }
+  _cursorHasMore = _cursor->hasMore();
+
+  return actuallySkipped;
+}
+
+std::tuple<ExecutorState, EnumerateCollectionStats, size_t, AqlCall> EnumerateCollectionExecutor::skipRowsRange(
+    AqlItemBlockInputRange& inputRange, AqlCall& call) {
+  AqlCall upstreamCall{};
+  EnumerateCollectionStats stats{};
+  bool offsetPhase = (call.getOffset() > 0);
+
+  TRI_ASSERT(_documentProducingFunctionContext.getAndResetNumScanned() == 0);
+  TRI_ASSERT(_documentProducingFunctionContext.getAndResetNumFiltered() == 0);
+
+  while (inputRange.hasDataRow() && call.shouldSkip()) {
+    uint64_t skipped = 0;
+
+    if (!_cursorHasMore) {
+      initializeNewRow(inputRange);
+    }
+
+    if (_cursorHasMore) {
+      TRI_ASSERT(_currentRow.isInitialized());
+      // if offset is > 0, we're in offset skip phase
+      if (offsetPhase) {
+        if (skipped < call.getOffset()) {
+          skipped += skipEntries(call.getOffset(), stats);
+        } else {
+          // we skipped enough in our offset phase
+          break;
+        }
+      } else {
+        // fullCount phase
+        _cursor->skipAll(skipped);
+        stats.incrScanned(skipped);
+        _documentProducingFunctionContext.getAndResetNumScanned();
```

Comment:
mhm do we actually need to hand in skipped. if we have getAndResetNumScanned()

I think the following is true here:
` _documentProducingFunctionContext.getAndResetNumScanned(); == skipped`
(on the first call of getAndReset) you might get away with this value only and do the call.didSkip() local to the skipIndex call.